src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -607,11 +608,24 @@ bool ProcessChunk(ReadOnlySpan<char> chunkDigits, ref Span<uint> currentBuffer)
                     int bufferSize = (totalDigitCount + MaxPartialDigits - 1) / MaxPartialDigits;
 
                     Span<uint> buffer = new uint[bufferSize];
+                    rentedBuffer = ArrayPool<int>.Shared.Rent(bufferSize);
+                    Span<uint> newBuffer = MemoryMarshal.Cast<int, uint>(rentedBuffer);
+                    newBuffer.Clear();
+
+                    // To ensure finally stored in newBuffer is the borrowed buffer.
+                    int blockSize = 1;
+                    do
+                    {
+                        Span<uint> tmp = buffer;
+                        buffer = newBuffer;
+                        newBuffer = tmp;
+                        blockSize *= 2;
+                    } while (blockSize < bufferSize);
```

Comment:
I don't understand why this is simply swapping the buffers back and forth until `blockSize` becomes greater than `bufferSize`

Reply:
This is necessary to ensure the value stored in newBuffer after exiting the main `while` loop is the borrowed buffer. (asserted in Line 754.) This swapping will happen as many times as the number of iterations of the main `while` loop.

This was thought to be necessary because after exiting `divide-and-conquer` algorithm, we use the function `MultiplyAdd`, which requires `buffer` to have been created from `arrayFromPool`.

https://github.com/dotnet/runtime/pull/55121/files/c3903db50a49c2202048b7428a17d2784f5e5f06..a9942c515df39408fed9b71225b321b19d94d1cc#diff-83081f6bd1123c4813c96a452ead0a065d670bbf44ed85023f772253eaf7d2c5R754

However, I noticed that `MultiplyAdd` does not require that it was created from `arrayFromPool`, as long as you are using it for `buffer`.  So this program will move as we expected even If we remove these steps. In the latest commit, these steps have been removed.